sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <paho-mqtt/MQTTClient.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <az_iot_provisioning_client.h>
+#include <az_result.h>
+#include <az_span.h>
+
+// Service information
+#define GLOBAL_PROVISIONING_ENDPOINT_DEFAULT "ssl://global.azure-devices-provisioning.net:8883"
+#define GLOBAL_PROVISIONING_ENDPOINT_ENV "AZ_IOT_GLOBAL_PROVISIONING_ENDPOINT"
+#define ID_SCOPE_ENV "AZ_IOT_ID_SCOPE"
+
+// Device information
+#define REGISTRATION_ID_ENV "AZ_IOT_REGISTRATION_ID"
+// AZ_IOT_DEVICE_X509_CERT_PEM_FILE is the path to a PEM file containing the device certificate and
+// key as well as any intermediate certificates chaining to an uploaded group certificate.
+
+#define DEVICE_X509_CERT_PEM_FILE "AZ_IOT_DEVICE_X509_CERT_PEM_FILE"
+
+static char global_provisioning_endpoint[256] = { 0 };
+static char id_scope[16] = { 0 };
+static char registration_id[256] = { 0 };
+static char x509_cert_pem_file[256] = { 0 };
+
+static az_iot_provisioning_client provisioning_client;
+static MQTTClient mqtt_client;
+
+static az_result read_configuration_entry(
+    const char* name,
+    const char* env_name,
+    char* default_value,
+    bool hide_value,
+    az_span buffer,
+    az_span* out_value)
+{
+  printf("%s = ", name);
+  char* env = getenv(env_name);
+
+  if (env != NULL)
+  {
+    printf("%s\n", hide_value ? "***" : env);
+    *out_value = az_span_init(az_span_ptr(buffer), 0, az_span_capacity(buffer));
+    AZ_RETURN_IF_FAILED(az_span_copy(*out_value, az_span_from_str(env), out_value));
+  }
+  else if (default_value != NULL)
+  {
+    printf("%s\n", default_value);
+    AZ_RETURN_IF_FAILED(az_span_copy(*out_value, az_span_from_str(default_value), out_value));
+  }
+  else
+  {
+    printf("(missing) Please set the %s environment variable.\n", env_name);
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+static az_result read_configuration()
+{
+  az_span endpoint_span = AZ_SPAN_LITERAL_FROM_BUFFER(global_provisioning_endpoint);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Global Device Endpoint",
+      GLOBAL_PROVISIONING_ENDPOINT_ENV,
+      GLOBAL_PROVISIONING_ENDPOINT_DEFAULT,
+      false,
+      endpoint_span,
+      &endpoint_span));
+
+  az_span id_scope_span = AZ_SPAN_LITERAL_FROM_BUFFER(id_scope);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "ID_Scope", ID_SCOPE_ENV, NULL, false, id_scope_span, &id_scope_span));
```

Comment:
>ID_Scope [](start = 7, length = 8)

macro, please do same for all

Reply:
If you think of a macro that'll define the spans, then definitely no.

Reply:
Literals, can they come from Macro?

---
In reply to: [404460738](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r404460738) [](ancestors = 404460738)